### PR TITLE
Tools for track creation and editing.

### PR DESCRIPTION
**Pull request by WhistleWiz**
This adds tools to create and edit tracks.
While most of the work can be done through a new editor window, all the functions are accessible without it.

### CODE REVIEW

File: MapifyEditor/Utils/Extensions.cs [hunk 1]
```diff
@@ -189,6 +230,75 @@ public static string PrettySceneName(this string sceneName)
...
+            return track.Curve[0].position.y == track.Curve[0].globalHandle2.y;
+        }
+
+        /// <summary>
+        /// Returns true if this track ends at grade of 0%.
+        /// </summary>
+        public static bool IsEndLevel(this Track track)
+        {
+            return track.Curve.Last().position.y == track.Curve.Last().globalHandle1.y;
+        }
```
Comment (Insprill) [inline MapifyEditor/Utils/Extensions.cs:254]: These don't appear to be used.

File: MapifyEditor/Utils/MathHelper.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                results[i - 1] = SampleBezier(
+                    curve[i - 1].position,
+                    curve[i - 1].globalHandle2,
+                    curve[i].globalHandle1,
+                    curve[i].position, samples);
+            }
+
+            return results;
+        }
```
Comment (Insprill) [inline MapifyEditor/Utils/MathHelper.cs:167]: Unused.

File: MapifyEditor/Utils/MathHelper.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+            result[samples] = result[0];
+
+            return result;
+        }
+
+        public static Vector3[] SampleBezier(Vector3 p0, Vector3 p1, Vector3 p2, Vector3 p3, int samples = 8)
+        {
+            Vector3[] results = new Vector3[samples];
+            float f;
```
Comment (Insprill) [inline MapifyEditor/Utils/MathHelper.cs:127]:
```suggestion

File: MapifyEditor/Utils/MathHelper.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+            for (int i = 0; i < samples; i++)
+            {
+                f = i / (float)(samples - 1);
```
Comment (Insprill) [inline MapifyEditor/Utils/MathHelper.cs:131]:
```suggestion
                float f = i / (float)(samples - 1);

File: MapifyEditor/Utils/MathHelper.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                return new Vector2(x, y);
+            }
+        }
+
+        public static Vector3[] SampleCircle(Vector3 center, float radius, int samples = 32)
+        {
+            Vector3[] result = new Vector3[samples + 1];
+            float angle;
+            float sin;
+            float cos;
```
Comment (Insprill) [inline MapifyEditor/Utils/MathHelper.cs:109]:
```suggestion

File: MapifyEditor/Utils/MathHelper.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+            for (int i = 0; i < samples; i++)
+            {
+                angle = (Tau * i) / samples;
+                sin = Mathf.Sin(angle);
+                cos = Mathf.Cos(angle);
```
Comment (Insprill) [inline MapifyEditor/Utils/MathHelper.cs:115]:
```suggestion
                float angle = (Tau * i) / samples;
                float sin = Mathf.Sin(angle);
                float cos = Mathf.Cos(angle);

File: MapifyEditor/Utils/MathHelper.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+            float determinant = a1 * b2 - a2 * b1;
+
+            if (determinant == 0)
+            {
+                // The lines are parallel. This is simplified
+                // by returning a pair of FLT_MAX
+                return new Vector2(float.PositiveInfinity, float.PositiveInfinity);
+            }
+            else
```
Comment (Insprill) [inline MapifyEditor/Utils/MathHelper.cs:96]: This else is redundant.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 8]
```diff
@@ -0,0 +1,1908 @@
...
+        // Settings.
+        private bool _showSettings = false;
+        private Color _forwardColour = Color.cyan;
+        private Color _backwardColour = Color.red;
+        private Color _newColour = Color.green;
+        private int _sampleCount = 8;
+
+        // Contents for the tool selection.
+        private readonly GUIContent[] _modeContents = new GUIContent[]
+        {
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:88]: Arrays created like this don't need an explicit type
```suggestion
        private readonly GUIContent[] _modeContents = {

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 9]
```diff
@@ -0,0 +1,1908 @@
...
+        private Vector3[][] _backwardLines = new Vector3[0][];
+
+        #endregion
+
+        #region PROPERTIES
+
+        private bool _isLeft => _orientation == TrackOrientation.Left;
+        public Track CurrentTrack => _selectedTracks.Length > 0 ? _selectedTracks[0] : null;
+        public BezierPoint CurrentPoint => _selectedPoints.Length > 0 ? _selectedPoints[0] : null;
+        public Switch CurrentSwitch => _isLeft ? LeftSwitch : RightSwitch;
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:132]: Unused.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 10]
```diff
@@ -0,0 +1,1908 @@
...
+
+                switch (_selectionType)
+                {
+                    case SelectionType.Track:
+                        DrawTrackSelection();
+                        break;
+                    case SelectionType.BezierPoint:
+                        DrawPointSelection();
+                        break;
+                    case SelectionType.None:
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:256]: This case is redundant.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 11]
```diff
@@ -0,0 +1,1908 @@
...
+                        EditorGUILayout.Space();
+                        break;
+                }
+
+                EditorGUILayout.Space();
+                EditorGUI.indentLevel--;
+            }
+
+            EditorGUILayout.EndFoldoutHeaderGroup();
+            //EditorGUILayout.LabelField("", (GUIStyle)"Tooltip", GUILayout.MaxHeight(1));
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:270]: Is this needed?

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 12]
```diff
@@ -0,0 +1,1908 @@
...
+
+            bool isSwitch = CurrentTrack.IsSwitch;
+
+            EditorGUILayout.BeginHorizontal();
+            EditorGUILayout.LabelField("Part of switch", isSwitch.ToString());
+
+            // If it's a switch add a button to swap between the 2 tracks.
+            if (isSwitch)
+            {
+                if (GUILayout.Button(new GUIContent("Swap tracks", "Swaps the selected track between the through and diverging tracks")))
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:289]: This if statement could be merged with the enclosing one to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 13]
```diff
@@ -0,0 +1,1908 @@
...
+            EditorGUILayout.LabelField(new GUIContent("Height difference", "Total change in height along the track"),
+                new GUIContent($"{CurrentTrack.GetHeightChange():F3}m"));
+            EditorGUILayout.LabelField(new GUIContent("Average grade", "Average grade of the track"),
+                new GUIContent($"{CurrentTrack.GetAverageGrade() * 100.0f:F2}%"));
+        }
+
+        private void DrawPointSelection()
+        {
+            EditorGUILayout.ObjectField(
+                new GUIContent($"Current point"),
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:343]:
```suggestion
                new GUIContent("Current point"),

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 14]
```diff
@@ -0,0 +1,1908 @@
...
+            }
+
+            EditorGUILayout.EndFoldoutHeaderGroup();
+        }
+
+        #region CREATION OPTIONS
+
+        private void DrawStraightOptions()
+        {
+            if (Require(TrackPrefab, "Track prefab"))
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:449]: This should be inverted to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 15]
```diff
@@ -0,0 +1,1908 @@
...
+
+                EditorGUILayout.LabelField(new GUIContent("Height dif new",
+                    "Height difference for new tracks"),
+                    new GUIContent($"{TrackToolsHelper.CalculateHeightDifference(0, _endGrade, _length):F3}m"));
+            }
+        }
+
+        private void DrawCurveOptions()
+        {
+            if (Require(TrackPrefab, "Track prefab"))
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:480]: This should be inverted to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 16]
```diff
@@ -0,0 +1,1908 @@
...
+                EditorGUILayout.EndHorizontal();
+
+                EditorGUILayout.LabelField(new GUIContent("Speed limit", "Estimated speed limit (shown on signs / closer estimate)"),
+                    new GUIContent($"{TrackToolsHelper.GetMaxSpeedForRadiusGame(_radius)}/{TrackToolsHelper.GetMaxSpeedForRadius(_radius):F1}km/h"));
+            }
+        }
+
+        private void DrawSwitchOptions()
+        {
+            if (Require(LeftSwitch, "Left switch prefab") &&
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:553]: This should be inverted to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 17]
```diff
@@ -0,0 +1,1908 @@
...
+                DrawOrientationGUI("Which side the diverging track turns to");
+                _connectingPoint = (SwitchPoint)EditorGUILayout.EnumPopup(new GUIContent("Connecting point",
+                    "Which of the 3 switch points should connect to the current track"),
+                    _connectingPoint);
+            }
+        }
+
+        private void DrawYardOptions()
+        {
+            if (Require(TrackPrefab, "Track prefab") &&
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:565]: This should be inverted to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 18]
```diff
@@ -0,0 +1,1908 @@
...
+                            new GUIContent($"{TrackToolsCreator.Previews.CachedYard.Value.TotalLength}m"));
+                        EditorGUILayout.LabelField(new GUIContent("Width", "The width of the yard, including loading gauge."),
+                            new GUIContent($"{totalTracks * _trackDistance + TrackToolsCreator.Previews.CachedYard.Value.LoadingGauge}m"));
+
+                        // .....
+                        //for (int i = 0; i < TrackToolsCreator.Previews.CachedYard.Value.SidingsLength.Length; i++)
+                        //{
+                        //    EditorGUILayout.LabelField(new GUIContent($"[{YardOptions.YardId}{YardOptions.StartTrackId + i}S]"),
+                        //        new GUIContent($"{TrackToolsCreator.Previews.CachedYard.Value.SidingsLength[i]}m"));
+                        //}
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:626]: Is this needed?

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 19]
```diff
@@ -0,0 +1,1908 @@
...
+                EditorGUILayout.LabelField("Total tracks", $"{totalTracks + 1}");
+                EditorGUILayout.LabelField("Track numbers", $"{_yardOptions.StartTrackId} to {_yardOptions.StartTrackId + totalTracks}");
+
+                if (_yardOptions.StartTrackId + totalTracks > 99)
+                {
+                    EditorGUILayout.HelpBox("Track number cannot exceed 99.", MessageType.Error);
+                }
+                _showYardCache = EditorGUILayout.Foldout(_showYardCache, new GUIContent("Yard sizes"));
+
+                if (_showYardCache)
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:609]: This should be inverted to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 20]
```diff
@@ -0,0 +1,1908 @@
...
+                    }
+
+                    EditorGUI.indentLevel--;
+                }
+            }
+        }
+
+        private void DrawTurntableOptions()
+        {
+            if (Require(TurntablePrefab, "Turntable prefab"))
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:636]: This should be inverted to reduce nesting.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 21]
```diff
@@ -0,0 +1,1908 @@
...
+                    EditorGUILayout.HelpBox("Not implemented yet!", MessageType.Warning);
+                    break;
+            }
+        }
+
+        #region SPECIAL OPTIONS
+
+        private void DrawSwitchCurveOptions()
+        {
+            if (Require(LeftSwitch, "Left switch prefab") &&
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:699]: Invert.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 22]
```diff
@@ -0,0 +1,1908 @@
...
+                {
+                    EditorGUILayout.HelpBox("The selected point has no connection with " +
+                        "the track, you should select one of the others.", MessageType.Warning);
+                }
+            }
+        }
+
+        private void DrawConnect2Options()
+        {
+            if (Require(TrackPrefab, "Track prefab"))
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:717]: Invert.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 23]
```diff
@@ -0,0 +1,1908 @@
...
+
+                _lengthMultiplier = EditorGUILayout.FloatField(
+                    new GUIContent("Length multiplier", "A multiplier that changes handle length, for smoother or tighter curves"),
+                    _lengthMultiplier);
+            }
+        }
+
+        private void DrawCrossoverOptions()
+        {
+            if (Require(TrackPrefab, "Track prefab") &&
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:764]: Invert.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 24]
```diff
@@ -0,0 +1,1908 @@
...
+                    "If the crossover is trailing into this direction"),
+                    _isTrailing);
+
+                DrawSwitchDistanceGUI();
+            }
+        }
+
+        private void DrawScissorsCrossoverOptions()
+        {
+            if (Require(TrackPrefab, "Track prefab") &&
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:783]: Invert.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 25]
```diff
@@ -0,0 +1,1908 @@
...
+                    "The distance between parallel tracks"),
+                    _trackDistance);
+
+                DrawSwitchDistanceGUI();
+            }
+        }
+
+        private void DrawDoubleSlipOptions()
+        {
+            if (Require(TrackPrefab, "Track prefab") &&
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:799]: Invert.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 26]
```diff
@@ -0,0 +1,1908 @@
...
+        #endregion
+
+        #endregion
+
+        private void DrawCreationButtons()
+        {
+            GUILayoutOption widthOption = GUILayout.Width(EditorGUIUtility.currentViewWidth * 0.2f);
+            GUILayoutOption smallWidth = GUILayout.Width(EditorGUIUtility.currentViewWidth * 0.1f);
+            // Tooltip to display, may change to explain why a button is disabled.
+            string tooltip = "";
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:820]:
```suggestion
            string tooltip;

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 27]
```diff
@@ -0,0 +1,1908 @@
...
+            menu.DropDown(rect);
+        }
+
+        private void DrawEditingFoldout()
+        {
+            GUI.backgroundColor *= 1.1f;
+
+            _showEditing = EditorGUILayout.BeginFoldoutHeaderGroup(_showEditing,
+                new GUIContent("Editing", "Ways to edit tracks after they've been created"),
+                null, null);
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:933]: Null is the default style and menuAction, no need to specify them explicitly.
```suggestion
            _showEditing = EditorGUILayout.BeginFoldoutHeaderGroup(_showEditing,
                new GUIContent("Editing", "Ways to edit tracks after they've been created"));

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 28]
```diff
@@ -0,0 +1,1908 @@
...
+                EditorGUI.indentLevel--;
+                EditorGUILayout.Space();
+            }
+
+            EditorGUILayout.EndFoldoutHeaderGroup();
+        }
+
+        private void PrefabFoldoutContextMenu(Rect rect)
+        {
+            var menu = new GenericMenu();
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1038]: Please use explicit types instead of var.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 29]
```diff
@@ -0,0 +1,1908 @@
...
+                EditorGUI.indentLevel--;
+                EditorGUILayout.Space();
+            }
+
+            EditorGUILayout.EndFoldoutHeaderGroup();
+        }
+
+        private void SettingsFoldoutContextMenu(Rect rect)
+        {
+            var menu = new GenericMenu();
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1087]: Here too.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 30]
```diff
@@ -0,0 +1,1908 @@
...
+                case SpecialTrack.ScissorsCrossover:
+                    SelectTrack(TrackToolsCreator.CreateScissorsCrossover(LeftSwitch, RightSwitch, TrackPrefab, _currentParent, attachPoint, handlePosition,
+                        _orientation, _trackDistance, _switchDistance, true)[3].ThroughTrack);
+                    break;
+                case SpecialTrack.DoubleSlip:
+                    SelectTrack(TrackToolsCreator.CreateDoubleSlip(LeftSwitch, RightSwitch, TrackPrefab, _currentParent, attachPoint, handlePosition,
+                        _orientation, _crossAngle, true)[2].ThroughTrack);
+                    break;
+                default:
+                    break;
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1201]: This is redundant.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 31]
```diff
@@ -0,0 +1,1908 @@
...
+            }
+        }
+
+        private void CreateConnect2(Vector3 attachPoint, Vector3 handlePosition)
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1205]: Neither parameter here is used. Is that a bug or can they be removed?

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 32]
```diff
@@ -0,0 +1,1908 @@
...
+                    break;
+                case SelectionType.None:
+                default:
+                    break;
+            }
+        }
+
+        public void DeleteTrack()
+        {
+            if (CurrentTrack)
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1229]: Invert.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 33]
```diff
@@ -0,0 +1,1908 @@
...
+            Vector3 pos;
+            Vector3 forward;
+
+            //if (_selectionType == SelectionType.BezierPoint)
+            //{
+            //    pos = CurrentPoint.position;
+            //    forward = CurrentPoint.position - CurrentPoint.globalHandle1;
+            //}
+            //else
+            {
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1433]: Is this needed?

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 34]
```diff
@@ -0,0 +1,1908 @@
...
+                return LeftSwitch;
+            }
+            else
+            {
+                return RightSwitch;
+            }
+        }
+
+        private void DrawHandles(SceneView scene)
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1778]: This should be broken up into multiple methods.

File: MapifyEditor/Tools/TrackToolsWindow.cs [hunk 35]
```diff
@@ -0,0 +1,1908 @@
...
+        }
+
+        // Check if an object is not null, draw an error box if it is.
+        private static bool Require(Object obj, string name)
+        {
+            if (obj)
+            {
+                return true;
+            }
+            else
```
Comment (Insprill) [inline MapifyEditor/Tools/TrackToolsWindow.cs:1898]: This else is redundant.

**Review by Insprill — CHANGES_REQUESTED** [review]
Thank you for the PR! Mostly just code-style things.

As a side note, are you willing to maintain this should any issues arise, or should any new features be added?

File: MapifyEditor/Tools/TrackToolsCreator.Previews.cs [hunk 36]
```diff
@@ -0,0 +1,498 @@
...
+
+                public YardCache(float totalLength, float[] sidingsLength, float loadingGauge)
+                {
+                    TotalLength = totalLength;
+                    SidingsLength = sidingsLength;
+                    LoadingGauge = loadingGauge;
+                }
+            }
+
+            internal static YardCache? CachedYard = null;
```
[the remark this answers is not in the document]
Comment (WhistleWiz) [reply MapifyEditor/Tools/TrackToolsCreator.Previews.cs:32]: Should the null not be assigned to explicitly state my intention?

[the remark this answers is not in the document]
Comment (Insprill) [reply MapifyEditor/Tools/TrackToolsCreator.Previews.cs:32]: Since it's a nullable type, it's implied that it's null unless explicitly set to something.

**Review by WhistleWiz — COMMENTED** [review]
Applied most of the changes, fixed bugs and added a feature regarding checks.
Didn't realise I had to press a button to get the review through...

**Comment by t0stiman** [comment]
Thanks for your work on this project @WhistleWiz ! 

One issue I found when generating a turntable:

**Comment by t0stiman** [comment]
The turntable bug I mentioned turned out to be my fault, ignore that.

**Comment by t0stiman** [comment]
I've been testing the track tools and it all seems to work well 👍

@Insprill any chance of a PR approval soon? :)

**Review by Insprill — APPROVED** [review]
LGTM. Thank you! 🥳

**Pull request merged**